torch/jit/annotations.py

```diff
@@ -0,0 +1,227 @@
+            raise RuntimeError("expected a single top-level function")
+        py_def = py_ast.body[0]
+        arg_types = [DynamicType() for _ in py_def.args.args]
+        if inspect.ismethod(fn):
+            arg_types = arg_types[1:]
+    else:
+        arg_types = [DynamicType()] * _n_arguments
+
+    return arg_types, ret_type
+
+
+def get_type_line(source):
+    """Tries to find the line containing a comment with the type annotation."""
+    lines = source.split('\n')
+
+    def strip_comment(line):
+        return line[:line.index('#') if '#' in line else None]
+
+    i = 0
+    while '):' not in strip_comment(lines[i]):
+        i += 1
+    i += 1
+
+    type_line = lines[i].strip()
+    if not type_line.startswith('# type:'):
+        return None
+    return type_line
+
+
+def split_type_line(type_line):
+    """Splits the comment with the type annotation into parts for argument and return types.
+
+    For example, for an input of:
+        # type: (Tensor, torch.Tensor) -> Tuple[Tensor, Tensor]
+
+    This function will return:
+        ("(Tensor, torch.Tensor)", "Tuple[Tensor, Tensor]")
+
+    """
+    start_offset = len('# type:')
+    try:
+        arrow_pos = type_line.index('->')
+    except ValueError:
+        raise RuntimeError("Syntax error in type annotation (cound't find `->`)")
+    return type_line[start_offset:arrow_pos].strip(), type_line[arrow_pos + 2:].strip()
+
+
+def try_real_annotations(fn):
+    """Tries to use the Py3.5+ annotation syntax to get the type."""
+    try:
+        sig = inspect.signature(fn)
+    except ValueError:
+        return None
+
+    all_annots = [sig.return_annotation] + [p.annotation for p in sig.parameters.values()]
+    if all(ann is sig.empty for ann in all_annots):
+        return None
+
+    def as_ann(ann):
+        # sig.empty is really annoying so convert it to None
+        return ann if ann is not sig.empty else None
+
+    param_types = [ann_to_type(as_ann(p.annotation))
+                   for p in sig.parameters.values()]
+    return_type = ann_to_type(as_ann(sig.return_annotation))
+    return param_types, return_type
+
+
+def ann_to_type(ann):
+    if ann is None:
+        return DynamicType()
+    elif ann is torch.Tensor:
+        return DynamicType()
+    elif is_tuple(ann):
+        return TupleType([ann_to_type(a) for a in ann.__args__])
+    raise ValueError("The only supported annotations kinds are Tensor and Tuple[...]")
+
+
+class Module(object):
+    def __init__(self, name, members):
+        self.name = name
+        self.members = members
+
+    def __getattr__(self, name):
+        try:
+            return self.members[name]
+        except KeyError:
+            raise RuntimeError("Module {} has no member called {}".format(self.name, name))
+
+
+env = {
+    'torch': Module('torch', {'Tensor': torch.Tensor}),
+    'Tensor': torch.Tensor,
+    'typing': Module('typing', {'Tuple': Tuple}),
+    'Tuple': Tuple,
+}
+
+
+def interpret_ann(expr):
```

Comment:
What is the reason to prefer using a mini-interpreter to using the Python interpreter here?